### DISCUSSION TITLE
Adding a new type of user address

### DISCUSSION THREAD

**Discussion by zernie** [solidusio/solidus#4043] [New Features or Ideas]
Currently, there are two types of user addresses: `billing` and `shipping`. For my marketplace (where users can own products) I'd like to add a new type, called `product`. However, it is hard to customize the logic because:
1. `Spree::UserAddressBook#save_in_address_book` needs to be overriden
  1.1 In here
https://github.com/solidusio/solidus/blob/e878076f2ed670d07654ab6293a16588743f2fa6/core/app/models/concerns/spree/user_address_book.rb#L122-L126
  1.2. and in here
https://github.com/solidusio/solidus/blob/e878076f2ed670d07654ab6293a16588743f2fa6/core/app/models/concerns/spree/user_address_book.rb#L113
     the problem with this line is that `user_addresses.mark_default` is defined only on `user_addresses` association:
     https://github.com/solidusio/solidus/blob/e878076f2ed670d07654ab6293a16588743f2fa6/core/app/models/concerns/spree/user_address_book.rb#L8-L14
      not on `Spree::UserAddress`, so you'd need to add a new custom method
2. `Spree::UserAddressBook#product_address=` needs to be added manually
3.  `Spree::UserAddressBook#product_address_attributes=` needs to be added manually
4. https://github.com/solidusio/solidus/blob/e878076f2ed670d07654ab6293a16588743f2fa6/core/app/models/concerns/spree/user_address_book.rb#L160-L164 needs to overriden
5. https://github.com/solidusio/solidus/blob/e878076f2ed670d07654ab6293a16588743f2fa6/core/app/models/concerns/spree/user_address_book.rb#L72-L91
needs to overriden
I might be missing something. But the point is: Could the logic of adding a new kind of address be generalized?

**Comment by jarednorman** [comment]
That's an interesting situation. I don't think I've ever seen one add more kinds of addresses like that.

**Reply by zernie** [reply]
@jarednorman haha. But do you think that I'm doing it correctly or should I take a different approach?

**Reply by jarednorman** [reply]
You've covered most of the main stuff I could think of. I think augmenting the address book system like this is the way to go, I'm just not sure this is a very common use case.

**Comment by kennyadsl** [comment]
Bill and Ship Addresses seem to have a lot of things in common. I'd be happy to review a PR where we try to generalize this concept, especially if this can be useful to create more address types as in your scenario.
